src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -4,26 +4,615 @@
+    Pose2d futureRobotPose =
+        new Pose2d(
+            originalSpeeds.vxMetersPerSecond * LOOP_TIME_S,
+            originalSpeeds.vyMetersPerSecond * LOOP_TIME_S,
+            Rotation2d.fromRadians(originalSpeeds.omegaRadiansPerSecond * LOOP_TIME_S));
+    Twist2d twistForPose = GeometryUtils.log(futureRobotPose);
+    ChassisSpeeds updatedSpeeds =
+        new ChassisSpeeds(
+            twistForPose.dx / LOOP_TIME_S,
+            twistForPose.dy / LOOP_TIME_S,
+            twistForPose.dtheta / LOOP_TIME_S);
+    return updatedSpeeds;
+  }
+
+  /** Keep modules in current position, don't drive */
+  public void setNoMove() {
+    Rotation2d frontLeftCurrRot = frontLeft.getPosition().angle;
+    Rotation2d frontRightCurrRot = frontRight.getPosition().angle;
+    Rotation2d rearLeftCurrRot = rearLeft.getPosition().angle;
+    Rotation2d rearRightCurrRot = rearRight.getPosition().angle;
+    frontLeft.setDesiredState(new SwerveModuleState(0, frontLeftCurrRot));
+    frontRight.setDesiredState(new SwerveModuleState(0, frontRightCurrRot));
+    rearLeft.setDesiredState(new SwerveModuleState(0, rearLeftCurrRot));
+    rearRight.setDesiredState(new SwerveModuleState(0, rearRightCurrRot));
+  }
+
+  /**
+   * Method to drive the robot using joystick info.
+   *
+   * @param xSpeed Desired speed of the robot in the x direction (forward), [-1,1].
+   * @param ySpeed Desired speed of the robot in the y direction (sideways), [-1,1].
+   * @param rot Desired angular rate of the robot, [-1,1].
+   * @param fieldRelative Whether the provided x and y speeds are relative to the field.
+   */
+  public void drive(double xSpeed, double ySpeed, double rot, boolean fieldRelative) {
+    if (isTimedMatch.getAsBoolean()
+        && DriverStation.isTeleop()
+        && DriverStation.getMatchTime() < 0.3) {
+      setX();
+      return;
+    }
+
+    // x, y, and rot are all being deadbanded from 0.1 to 0.0, so checking if
+    // they're equal to 0
+    // does account for controller deadzones.
+    if (xSpeed == 0 && ySpeed == 0 && rot == 0) {
+      setNoMove();
+      return;
+    }
+
+    // Adjust input based on max speed
+    xSpeed *= DriveConstants.MAX_SPEED_METERS_PER_SECOND;
+    ySpeed *= DriveConstants.MAX_SPEED_METERS_PER_SECOND;
+    rot *= DriveConstants.MAX_ANGULAR_SPEED_RAD_PER_SECONDS;
+
+    xSpeed *= throttleMultiplier;
+    ySpeed *= throttleMultiplier;
+    rot *= throttleMultiplier;
+
+    ChassisSpeeds desiredChassisSpeeds =
+        fieldRelative
+            ? ChassisSpeeds.fromFieldRelativeSpeeds(
+                xSpeed, ySpeed, rot, Rotation2d.fromDegrees(gyro.getYaw()))
+            : new ChassisSpeeds(xSpeed, ySpeed, rot);
+
+    desiredChassisSpeeds = correctForDynamics(desiredChassisSpeeds);
+    lastSetChassisSpeeds = desiredChassisSpeeds;
+
+    var swerveModuleStates =
+        DriveConstants.DRIVE_KINEMATICS.toSwerveModuleStates(desiredChassisSpeeds);
+
+    SwerveDriveKinematics.desaturateWheelSpeeds(
```

Comment:
Are we already doing this desaturation in the module?

Reply:
Might not be

Reply:
Nope, we aren't

Reply:
Could you explain what this does? And maybe add a comment

Reply:
@avakimc Here's the description:

<img width="785" alt="image" src="https://github.com/cachemoney8096/2023-minibot/assets/22648713/c94923b7-5f85-424f-a812-5becb255df51">

Basically, because of some swerve math I don't really understand, the calculated speed from the modules might end up being more than the max physical speed. This function keeps the *ratio* of the speeds but it limits them to all be physically possible.